### PR TITLE
Add a BufferedReader and allow BufferedWriter to handle partial writes and errors after some data was written.

### PR DESCRIPTION
**Pull request by klondi**
The commits should already describe much of what is going on but in summary.

Fix a missing include in py/mpconfig.h
Update BufferedWriter to handle partial writes and errors after some data is written.
Add a BufferedReader class handling also partial reads and errors after some data is read.

This is my first time sending code to micropython so I hope I did not make any mistakes.

(Also in case it is necessary, yes the code is mine, it is owned by me and not by any of my employers and I agree to release it under the MIT License).

### CODE REVIEW

**Comment by klondi** [comment]
I have fixed the spelling error, I am unsure what should I do with the Signed-Of-By headers.

File: py/misc.h [hunk 1]
```diff
@@ -33,6 +33,8 @@
 #include <stdbool.h>
 #include <stdint.h>
 #include <stddef.h>
+// Needed for NORETURN
+#include "py/mpconfig.h"
```
Comment (stinos) [inline py/misc.h:37]: Don't think this is explicitly documented, but see the rest of the code (grep for all occurrences of misch.h): the idea is that any file needing to include py/misc.h first includes py/mpconfig.h, so this change should not be made.

File: py/modio.c [hunk 2]
```diff
@@ -123,61 +126,100 @@ STATIC mp_obj_t bufwriter_make_new(const mp_obj_type_t *type, size_t n_args, siz
...
     return o;
 }
 
+#endif // MICROPY_PY_IO_BUFFEREDWRITER || MICROPY_PY_IO_BUFFEREDREADER
+
+#if MICROPY_PY_IO_BUFFEREDWRITER
+// Writes out the data stored in the buffer so far
+STATIC int bufwriter_do_write(mp_obj_bufwriter_t *self) {
+    int rv = 0;
+    // This cannot return 0 without an error
```
Comment (stinos) [inline py/modio.c:139]: Can you elaborate on what you mean with this? Like: why would it be an issue for this bit of code if it were 0?

File: py/modio.c [hunk 3]
```diff
@@ -123,61 +126,100 @@ STATIC mp_obj_t bufwriter_make_new(const mp_obj_type_t *type, size_t n_args, siz
...
+        memmove(self->buf, self->buf + out_sz, self->len);
+    }
+    return rv;
+}
+
 STATIC mp_uint_t bufwriter_write(mp_obj_t self_in, const void *buf, mp_uint_t size, int *errcode) {
     mp_obj_bufwriter_t *self = MP_OBJ_TO_PTR(self_in);
 
     mp_uint_t org_size = size;
+    // Alloc should always remain the same so cache it.
```
Comment (stinos) [inline py/modio.c:154]: Minor, but here the sentence ends with a dot (imo best), but not the other comments. Also the comment below is after its statement instead of on the line above it.

**Comment by stinos** [comment]
Would it be possible to add tests? The code is complex enogh to warrant testing all possible edge cases imo.

**Comment by projectgus** [comment]
This is an automated heads-up that we've just merged a Pull Request
that removes the STATIC macro from MicroPython's C API.

See https://github.com/micropython/micropython/pull/13763

A search suggests this PR might apply the STATIC macro to some C code. If it
does, then next time you rebase the PR (or merge from master) then you should
please replace all the `STATIC` keywords with `static`.

Although this is an automated message, feel free to @-reply to me directly if
you have any questions about this.

**Comment by DvdGiessen** [comment]
I've rebased them to match all the changes that happened in the `master` branch since this PR was made, and applied the review suggestion to import `mpconfig.h` directly in `modio.h` (we logically need it for `MICROPY_PY_IO_BUFFEREDWRITER || MICROPY_PY_IO_BUFFEREDREADER` anyway). Please feel free to copy these over: https://github.com/DvdGiessen/micropython/compare/master...bufferedreaderwriter

Tested it and it works well for significantly speeding up DeflateIO when it's doing many small reads from a not so efficient custom stream. I'd love to see this contribution included.

I'll open a small PR for the `Signed-Off-By` breaking the rules. Long names should be allowed.

**Comment by projectgus** [comment]
> I've rebased them to match all the changes that happened in the `master` branch since this PR was made, and applied the review suggestion to import `mpconfig.h` directly in `modio.h` (we logically need it for `MICROPY_PY_IO_BUFFEREDWRITER || MICROPY_PY_IO_BUFFEREDREADER` anyway). Please feel free to copy these over: [DvdGiessen/micropython@master...bufferedreaderwriter](https://github.com/DvdGiessen/micropython/compare/master...bufferedreaderwriter)

Thanks for picking this up, @DvdGiessen. Unsure how best to handle it, as the PR's been open for a long time. However I've taken the liberty of updating @klondi's branch in this PR with your changes. If we end up going back and forth more than a little then it might be best to open a new PR instead, though.

**Comment by projectgus** [comment]
To keep code coverage happy (and to give us the confidence of test coverage), this probably needs some unit tests. The `coverage` configuration of the unix port is built at "everything" level so it should have `io.BufferedReader` already.

It seems to me that the "standard" config of the unix port might benefit from having `BufferedReader` and `BufferedWriter` manually enabled, as well.

**Comment by dpgeorge** [comment]
IMO this is a worthy feature to have (although at what level to enable it is a separate question).  But it will need:
- tests, full test coverage
- docs
- check API compatibility with CPython

**Comment by dpgeorge** [comment]
Closing due to inactivity.
